File: READY_REPO.cpp
#include <iostream>
#include <algorithm>
#include <filesystem>
#include <vector>
#include <fstream>

namespace fs = std::filesystem;

// functions pre-declaration

bool is_source_code_file(std::string file_name);
bool good_name(std::string name);
bool update_documentation(std::vector<std::string> folders, std::vector<std::pair<std::string, std::string>> files);

// main function
int main()
{
    try
    {
        // intro
        std::cout << "read_repository is a special program specifically written for this repository to auto generate DOCUMENTATION.md\n\
        * Make sure you are running this program just before commit a change.\n\
        * Do not change/modify this program.\n\
        * Please read the CONTRIBUTE.md to insure you are following all \n\
        the rules for contribute in this project\n\
        (https://github.com/mursalatul/code-park/blob/master/CONTRIBUTE.md).\n\
        * You will see a positive message if all goes well.\n\
        \nProcess started\n";
        // all folder list
        std::vector<std::string> folders;
        // all source code file list
        std::vector<std::pair<std::string, std::string>> files; // here first = file name, second = directory name

        // taking the currect path of all the folders
        std::string path = fs::current_path().string();

        // iterate all the folder and source codes and store them in the above vectors
        for (const auto &entry : fs::directory_iterator(path))
        {
            // folder
            if (fs::is_directory(entry.status()))
            {
                // check for valid folder name store it
                if (good_name(entry.path().filename()))
                {
                    folders.push_back(entry.path().filename());

                    // go to the valid folder and store files in the folder
                    std::string path_in_folder = entry.path().string();
                    for (const auto &entry2 : fs::directory_iterator(path_in_folder))
                    {
                        if (fs::is_regular_file(entry2.status()))
                        {
                            // check for valid file name and store it
                            if (is_source_code_file(entry2.path().filename()))
                            {
                                files.push_back({entry2.path().filename(), entry.path().filename()});
                            }
                        }
                    }
                }
            }
        }

        // sorting folder and file name for better view in the documentation.
        std::sort(folders.begin(), folders.end());
        std::sort(files.begin(), files.end());

        // updating documentation
        bool status = update_documentation(folders, files);
        if (status)
        {
            std::cout << "Process completed!\nDOCUMENATION.md has been successfully updated.\n";
            std::cout << "\nCheck DOCUMENTATION.md for changes.\nHappy Coding!\n";
        }
        else
            std::cout << "Process Failed!\nDOCUMENATION.md update failed.\n";
    }
    catch (const fs::filesystem_error &ex)
    {
        std::cerr << "Error accessing directory: " << ex.what() << std::endl;
    }
    return 0;
}

// function definition

/*
check if the file is a valid source code file

Parameters:
file name (string)

Return:
bool: true if file is a correct source code file, else false
*/
bool is_source_code_file(std::string file_name)
{
    // check if the file is a valid file
    std::string language_extension[] = {".c", ".cpp", ".py", ".java"};
    for (std::string extension : language_extension)
    {
        size_t extension_index = file_name.find(extension);
        if (extension_index != std::string::npos)
        {
            //
            std::string name_with_out_extension = file_name.substr(0, extension_index);
            // check if the name is good name or not
            if (good_name(name_with_out_extension))
                return true;
        }
    }
    return false;
}

/*
check if a name using proper format or not.
Example, a name (file/folder) can not contain any special char

Parameter:
string: file or folder name to check

Return:
bool: true if the name format is good, else false.
*/
bool good_name(std::string name)
{
    // check if the name contain any special character
    std::string exclude_char_list[] = {".", "-", "!", "@", "#", "$", "%", "^", "&", "*", "(", ")", "+", "=", "~", "`", "{", "}", "|", "\\", ":", ";", "'", "\"", ",", ".", "?", "/"};
    for (std::string exclude_char : exclude_char_list)
    {
        // if special char found return false
        size_t index = name.find(exclude_char);
        if (index != std::string::npos)
            return false;
    }

    // check if the  name contail any upper case letter
    for (char name_letter : name)
    {
        // return false if upper case found
        if (name_letter >= 'A' && name_letter <= 'Z')
            return false;
    }

    // name is good
    return true;
}

/*
update the DOCUMENTATION.md with updated new folders and files

Parameters:
vector 1: folder
vector 2: file

Return:
bool: true if the update process complete, else false
*/
bool update_documentation(std::vector<std::string> folders, std::vector<std::pair<std::string, std::string>> files)
{
    std::string filename = "DOCUMENTATION.md";
    // Open the file in reading and writing mode
    std::fstream fileStream(filename, std::ios::out);

    // store status
    bool status = false;

    // Write lines to the file
    if (fileStream.is_open())
    {
        // adding heading
        fileStream << "# Code Park Documentation\nThis document encompasses all the current subjects within Code Park.\
        \n\n## Folders\n";

        // adding folders
        for (std::string folder_name : folders)
            fileStream << "- [" << folder_name << "](/" << folder_name << ")\n";

        // adding files
        fileStream << "\n## Files\n";
        for (std::pair<std::string, std::string> f : files)
            fileStream << "- [" << f.first << "](/" << f.second << "/" << f.first << ")\n"; // file.first -> file name, file.second -> folder name

        // Close the file
        fileStream.close();

        status = true;
    }
    else
    {
        std::cerr << "Error opening file " << filename << " for writing." << std::endl;
        status = false;
    }
    return status;
}

File: READY_REPO.py
import os

class CreateDocumentation:
    def writedoc(self, folder_file: dict):
        with open("DOCUMENTATION.md", "w") as doc:
            doc.write("# Code Park Documentation\n\
This documentation provides a front view of the content presented in the repository. Here, you can see the folders and the files with a tree-view structure.\n\n")

            doc.write("## Content\n")

            # sorting the folders before adding to doc
            folder_file = dict(sorted(folder_file.items()))

            # getting folder name and file list
            for folder, file in folder_file.items():
                # writing folder name to doc
                doc.write(f"- [{folder}](/{folder})\n")
                # sorting the files for organized view in doc
                file.sort()
                # adding the file names
                for files in file:
                    doc.write(f"  - [{files}](/{files})\n")

class Files:
    # this char will not accepted inside a class
    RESTRICTED_CHAR = [".", "-", "!", "@", "#", "$", "%", "^", "&", "*", 
                    "(", ")", "+", "=", "~", "`", "{", "}", "|", "\\",
                    ":", ";", "'", '"', ",", ".", "?", "/", " "]
    
    # special cases to restrict
    # this folders will not be considered as regular folder to add in DOCUMENTATION.md
    SPECIAL_FILES_TO_IGNORE = ["LICENSE", "venv", "virtual_env", "env", "environment"]
    
    # this languages only will be accespted to commit
    LENGUAGE_EXTENTION = [".cpp", ".c", ".py", ".java"]

    def get_all_valid_folder_files_dict(self):
        # store all the data
        """
        structure:
        {
            folder1 : [file1, file2, .....],
            folder2 : [file1, file2, .....],
            .
            .
            .
        """
        all_files = {}
        cwd = os.getcwd()
        for folder_name in os.listdir(cwd):
            # works only for directory
            if os.path.isdir(os.path.join(cwd, folder_name)):
                # checking if folder_name contain any RESTRICTED_CHAR
                folder_status = self.goodfolderformat(folder_name)["status"]
                if folder_status == "ignore_folder" or folder_name in self.SPECIAL_FILES_TO_IGNORE:
                    pass
                elif folder_status == "1":
                    # store all the file names
                    file_names = []

                    # going inside the folder to get all the files
                    inside_folder_name = os.path.join(cwd, folder_name)

                    for file in os.listdir(inside_folder_name):
                        file_path = os.path.join(inside_folder_name, file)
                        if os.path.isfile(file_path):
                            # checking if the file is a valid file
                            file_status = self.good_file_format(file)["status"]
                            if file_status == '1' and file not in self.SPECIAL_FILES_TO_IGNORE:
                                file_names.append(file)
                            else:
                                # if wrong file found
                                return {"ERROR": file_status,
                                        "target" : "file",
                                        "name" : file}

                    all_files[folder_name] = file_names
                else:
                    # if wrong formatted folder found
                    return {"ERROR" : folder_status,
                            "target" : "folder",
                            "name" : folder_name}
        return (all_files)

    def isgoodname(self, name: str) -> dict:
        """
        check if the name contain any restricted character or not
        
        Parameters:
        - (str) name to check

        Returns
        - (dict) 
        "status" : 1 if not found any restricted character
        "status" : "restricted character" if restricted character found
        """
        for character_in_name in list(name): # converting str -> to iterater
            if character_in_name in self.RESTRICTED_CHAR:
                status =  {"status" : ""}
                # finding the restricted character and adding it in dict
                for c in self.RESTRICTED_CHAR:
                    if character_in_name == c:
                        status["status"] = c
                        return status
        return {"status": "1"}
    
    def good_file_format(self, file_name: str) -> dict:
        status = False
        for lang_ext in self.LENGUAGE_EXTENTION:
            if file_name.find(lang_ext) != -1:
                status = True
                break
        # if not from the LENGUAGE_EXTENTION
        if status == False:
            return {"status" : f"currectly only accepting codes on {self.LENGUAGE_EXTENTION}"}
        else:
            file_name_with_out_ext = []
            for c in list(file_name):
                if c != '.':
                    file_name_with_out_ext.append(c)
                else:
                    break
            return self.isgoodname("".join(file_name_with_out_ext))

    def goodfolderformat(self, folder_name: str) -> dict:
        if folder_name[:1] == ".":
            return {"status" : "ignore_folder"}
        else:
            return self.isgoodname(folder_name)


def main():
    # welcome message
    print("Preparing the repository for commit.".center(50, "*"))
    print("\nFolder/File naming format checking . . . . . . . .")
    ff = Files()
    data = (ff.get_all_valid_folder_files_dict())
    if "ERROR" in data:
        # show relevent error message
        # if error with the folder name
        if data["target"] == "folder":
            # show error message and relavent error(RESTRICTED_CHAR)
            print(f"Wrong folder naming. \nYou have used \"", data["ERROR"], "\" in your folder name", sep="")
            # show the folder name with error
            print("->", data["name"])
        # if error with file name
        else:
            # show error message and relavent error(RESTRICTED_CHAR)
            print(f"Wrong file naming. \nYou have used \"", data["ERROR"], "\" in your folder name", sep="")
            # show the file name with error
            print("->", data["name"])
        print("\nChecking process ended with error!")
    else:
        print("Checking process ended. All Folder/File names are seems good!\n")
        doc = CreateDocumentation()
        doc.writedoc(data)
        print("DOCUMENTATION.md updated successful.\nYou can commit your updates now\n")

        # ending message
        print("Some key points to remember:")
        print("* make sure you are running this program just before the commit.\n\
* do not modify this program. this is build to reduce your effort.\n\
* if you find any bug/suggestion feel free to submit an issue at\n\
https://github.com/mursalatul/code-park/issues/new \n\
* see the CONTRUBUTION.md(https://github.com/mursalatul/code-park/blob/master/CONTRIBUTE.md)\n\
if you are confused about contributing in this project.\n")

if __name__ == '__main__':
    main()

